pkg/execution/executor/executor.go

```diff
@@ -568,74 +563,21 @@ func (e *executor) Execute(ctx context.Context, id state.Identifier, item queue.
 		))
 	defer span.End()
 
-	if md.Status == enums.RunStatusCancelled {
-		return nil, state.ErrFunctionCancelled
-	}
-
-	if md.Status == enums.RunStatusScheduled {
-		if err := e.sm.SetStatus(ctx, id, enums.RunStatusRunning); err != nil {
-			return nil, err
-		}
-
-		// Reload the metadata to ensure we get all changes that happen when
-		// setting the status.
-		//
-		// TODO: Refactor the state store so that we don't need to reload when
-		// setting the status
-		s, err := e.sm.Load(ctx, id.RunID)
-		if err != nil {
-			return nil, err
-		}
-		md = s.Metadata()
+	f, err := e.fl.LoadFunction(ctx, id)
+	if err != nil {
+		return nil, fmt.Errorf("error loading function for run: %w", err)
 	}
 
-	if e.steplimit != 0 && len(s.Actions()) >= int(e.steplimit) {
-		// Update this function's state to overflowed, if running.
-		if md.Status == enums.RunStatusRunning {
-			// XXX: Update error to failed, set error message
-			if err := e.sm.SetStatus(ctx, id, enums.RunStatusFailed); err != nil {
-				return nil, err
-			}
-
-			// Create a new driver response to map as the function finished error.
-			resp := state.DriverResponse{}
-			resp.SetError(state.ErrFunctionOverflowed)
-			resp.SetFinal()
-
-			if err := e.runFinishHandler(ctx, id, s, resp); err != nil {
-				logger.From(ctx).Error().Err(err).Msg("error running finish handler")
-			}
-
-			for _, e := range e.lifecycles {
-				go e.OnFunctionFinished(context.WithoutCancel(ctx), id, item, resp, s)
-			}
-		}
-		return nil, state.ErrFunctionOverflowed
+	// Validate that the run can execute.
+	v := newRunValidator(item, s, f, e)
```

Comment:
The mess of `if` guards before running steps really got to me.  This is the start of a cleaner approach and can be continued as we refactor the executor.